ServiceBus EventSource remove allocations and boxing

danielmarbach opened the pull request:
Similar to https://github.com/Azure/azure-sdk-for-net/pull/26989 but for ServiceBus to make @JoshLove-msft a happy person too.

# Contributing to the Azure SDK

Please see our [CONTRIBUTING.md](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md) if you are not familiar with contributing to this repository or have questions.

For specific information about pull request etiquette and best practices, see [this section](https://github.com/Azure/azure-sdk-for-net/blob/main/CONTRIBUTING.md#pull-request-etiquette-and-best-practices).

danielmarbach committed:
ServiceBus EventSource remove allocations and boxing

jsquire commented:
/azp run net - servicebus - tests

jsquire commented:
The test `LogsTransactionEvents` seems to be the source of failures, though it appears to be flaky rather than outright failing.   Might be a timing issue or may be something more involved.  I haven't had a chance to dig into it at all.

danielmarbach commented:
I can also look into it sometime next week. I'm mostly AFK enjoying carnival 🙃

danielmarbach committed:
IsEnabled

danielmarbach commented:
Pushed an update

danielmarbach commented:
Is it possible to rerun the tests here to see if things are still flaky?

jsquire commented:
/azp run net - servicebus - tests

danielmarbach committed:
Make event method ProcessorMessageHandlerExceptionCore not unsafe

danielmarbach commented:
Eyeballing it again I think the order of parameters, types as well as the sizes of the parameters should match what we had before. I can explain why it is failing. I pushed now an update the removes the `unsafe` on one of the Event methods to make sure we have nothing `unsafe` on the event methods. Other than that I'm a bit clueless what it could be. I need to probably setup the test environment at some point on my machine to get more insights...

JoshLove-msft commented:
/azp run net - servicebus - tests

jsquire commented:
> Eyeballing it again I think the order of parameters, types as well as the sizes of the parameters should match what we had before. I can explain why it is failing. I pushed now an update the removes the `unsafe` on one of the Event methods to make sure we have nothing `unsafe` on the event methods. Other than that I'm a bit clueless what it could be. I need to probably setup the test environment at some point on my machine to get more insights...

Agreed.   I've revisited this a couple of times in the last two weeks, and I am at a loss.   The results are consistent, though, and show that something in `LogsTransactionEvents` gets upset.  I wonder if this maybe has something to do with [TestEventListener](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core.TestFramework/src/TestEventListener.cs) rather than the event source itself?   Or maybe the `Delay` used in the test?   I haven't been able to dig into those areas yet.

danielmarbach committed:
Mad scientist stuff

danielmarbach commented:
Narrator: It was a bool who made him not sleep.

danielmarbach commented:
This simple repro helped quite a bit

``` 
// See https://aka.ms/new-console-template for more information
using System.Diagnostics.Tracing;
using System.Runtime.CompilerServices;

SomeEventSourceListener listener = new SomeEventSourceListener();

SomeEventSource.Log.TransactionDischarged("TransactionId", " AmqpTransactionId", true);

Console.ReadLine();


sealed class SomeEventSourceListener : EventListener
{
    protected override void OnEventSourceCreated(EventSource eventSource)
    {
        if (eventSource.Name.Equals("Some-EventSource"))
        {
            EnableEvents(eventSource, EventLevel.Verbose, EventKeywords.All);
        }
    }

    // from https://blogs.msdn.microsoft.com/dotnet/2018/12/04/announcing-net-core-2-2/
    // Called whenever an event is written.
    protected override void OnEventWritten(EventWrittenEventArgs eventData)
    {
        // Write the contents of the event to the console.
        Console.WriteLine($"ThreadID = {eventData.OSThreadId} ID = {eventData.EventId} Name = {eventData.EventName}");
        for (int i = 0; i < eventData.Payload.Count; i++)
        {
            string payloadString = eventData.Payload[i] != null ? eventData.Payload[i].ToString() : string.Empty;
            Console.WriteLine($"    Name = \"{eventData.PayloadNames[i]}\" Value = \"{payloadString}\"");
        }
        Console.WriteLine("\n");
    }
}

[EventSource(Name = EventSourceName)]
internal class SomeEventSource : EventSource
{
    private const string EventSourceName = "Some-EventSource";

    internal const int TransactionDischargedEvent = 91;

    protected SomeEventSource()
    {
    }

    public static SomeEventSource Log { get; } = new SomeEventSource();

    [Event(TransactionDischargedEvent, Level = EventLevel.Informational, Message = "AmqpTransactionDischarged for LocalTransactionId: {0} AmqpTransactionId: {1} Rollback: {2}.")]
    public void TransactionDischarged(string transactionId, string amqpTransactionId, bool rollback)
    {
        //WriteEvent(TransactionDischargedEvent, transactionId, amqpTransactionId, rollback);
        TransactionDischargedCore(TransactionDischargedEvent, transactionId, amqpTransactionId, rollback);
    }

    [NonEvent]
    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    private unsafe void TransactionDischargedCore(int eventId, string transactionId, string amqpTransactionId, bool rollback)
    {
        fixed (char* transactionIdPtr = transactionId)
        fixed (char* amqpTransactionIdPtr = amqpTransactionId)
        {
            var eventPayload = stackalloc EventData[3];

            eventPayload[0].Size = (transactionId.Length + 1) * sizeof(char);
            eventPayload[0].DataPointer = (IntPtr)transactionIdPtr;

            eventPayload[1].Size = (amqpTransactionId.Length + 1) * sizeof(char);
            eventPayload[1].DataPointer = (IntPtr)amqpTransactionIdPtr;

            var rollbackInt = Convert.ToInt32(rollback);
            eventPayload[2].Size = Unsafe.SizeOf<int>();
            eventPayload[2].DataPointer = (IntPtr) Unsafe.AsPointer(ref rollbackInt);

            WriteEventCore(eventId, 3, eventPayload);
        }
    }
}
```

danielmarbach commented:
This should now pass (famous last words)

JoshLove-msft commented on sdk/servicebus/Azure.Messaging.ServiceBus/src/Diagnostics/ServiceBusEventSource.cs:
```diff
@@ -1509,7 +1590,34 @@ public void TransactionDischarged(string localTransactionId, ArraySegment<byte>
+                eventPayload[1].DataPointer = (IntPtr)amqpTransactionIdPtr;
+
+                // bool maps to "win:Boolean", a 4-byte boolean
+                var rollbackInt = Convert.ToInt32(rollback);
```
There are other events that use bool parameter. Why do we only need to update this one?

danielmarbach replied:
Because I was so excited that I found the root cause that I forgot to search for other occurrences of this problem. Will update. Thanks for your patience 😅

JoshLove-msft replied:
Hehe no worries - I wasn't sure if we were missing coverage for that one, but I guess we didn't run the tests after your last commit. Let me try that now to verify that we have coverage.

JoshLove-msft replied:
Probably a more efficient way for me to determine this would be to just look in our tests 😆 
It doesn't look like we have coverage for the LinkStateLost event source event. I think this one would be difficult to test as a live event source test, but hopefully we can add a mock test for it in EventSourceTests.

danielmarbach replied:
It uses a test event listener which is similar to my repro code. It would have caught it. I can see what I can do at some later point

danielmarbach replied:
I briefly looked at the EventSourceTests and since they use the higher level API surface to simulate the client into the cases that triggers the event log entry it would be quite involved for me to understand how to mock the client into a situation where we get a linked lost within the right state. So if any of you can quickly bang this out I wouldn't mind direct pushes to the branch. Otherwise it is going to take a while for me to get to this.

JoshLove-msft replied:
Appreciate your patience @danielmarbach - I'll try to look into adding a test sometime this week.

danielmarbach replied:
I had a quick look, started adding an internal property to the `AmqpReceiver` that would make it possible to override the event source. Then I ran into the rabbit hole of somehow having to mock away the `ReceivingAmqpLink` which I couldn't find out with reasonable effort at this stage. 

Then I also realized we would need to figure out where to place the tests because the `EventSourceTests` currently test from the higher level while the event log in question was for the lower level AMQP connection events which seems we haven't really tried to unit test. That in itself is probably a lot of effort for me to get working. Given there is no real test coverage of these paths I'm inclined to say that improvement could be done separately from this PR (but of course I'm biased to get this in :) )

JoshLove-msft replied:
So I tried closing the underlying receive link but that ends up resulting in a lock lost exception rather than the OperationCanceledException. Because we can't mock the underlying AMQP methods it seems like it will be hard to reliably test this. I think we can go ahead with this as is, since as you mentioned, we didn't have coverage for this event before either.

JoshLove-msft commented:
/azp run net - servicebus - tests

danielmarbach commented on sdk/servicebus/Azure.Messaging.ServiceBus/src/Diagnostics/ServiceBusEventSource.cs:
```diff
@@ -1509,7 +1590,34 @@ public void TransactionDischarged(string localTransactionId, ArraySegment<byte>
+                eventPayload[1].DataPointer = (IntPtr)amqpTransactionIdPtr;
+
+                // bool maps to "win:Boolean", a 4-byte boolean
+                var rollbackInt = Convert.ToInt32(rollback);
+                eventPayload[2].Size = Unsafe.SizeOf<int>();
+                eventPayload[2].DataPointer = (IntPtr)Unsafe.AsPointer(ref rollbackInt);
```
```suggestion
                var rollbackInt = rollback ? bool.True : bool.False;
                eventPayload[2].Size = Unsafe.SizeOf<int>();
                eventPayload[2].DataPointer = (IntPtr)Unsafe.AsPointer(ref rollbackInt);
```

danielmarbach commented on sdk/servicebus/Azure.Messaging.ServiceBus/src/Diagnostics/ServiceBusEventSource.cs:
```diff
@@ -1002,7 +1050,37 @@ public virtual void LinkStateLost(string identifier, string receiveLinkName, str
+                eventPayload[2].Size = (receiveLinkState.Length + 1) * sizeof(char);
+                eventPayload[2].DataPointer = (IntPtr)receiveLinkStatePtr;
+
+                eventPayload[3].Size = Unsafe.SizeOf<bool>();
+                eventPayload[3].DataPointer = (IntPtr)Unsafe.AsPointer(ref isSessionReceiver);
```
```suggestion
                var isSessionReceiverInt = isSessionReceiver ? bool.True : bool.False;
                eventPayload[3].Size = Unsafe.SizeOf<int>();
                eventPayload[3].DataPointer = (IntPtr)Unsafe.AsPointer(ref isSessionReceiverInt);
```

danielmarbach committed:
Adjust boolean conversion and fix left out part

danielmarbach commented on sdk/servicebus/Azure.Messaging.ServiceBus/src/Diagnostics/ServiceBusEventSource.cs:
```diff
@@ -1002,7 +1050,38 @@ public virtual void LinkStateLost(string identifier, string receiveLinkName, str
+                eventPayload[2].Size = (receiveLinkState.Length + 1) * sizeof(char);
+                eventPayload[2].DataPointer = (IntPtr)receiveLinkStatePtr;
+
+                var isSessionReceiverInt = isSessionReceiver ? bool.True : bool.False;
```
```suggestion
                // bool maps to "win:Boolean", a 4-byte boolean
                var isSessionReceiverInt = isSessionReceiver ? bool.True : bool.False;
```

danielmarbach committed:
Comment on the conversion

danielmarbach commented on sdk/servicebus/Azure.Messaging.ServiceBus/src/Diagnostics/ServiceBusEventSource.cs:
```diff
@@ -1002,7 +1050,39 @@ public virtual void LinkStateLost(string identifier, string receiveLinkName, str
+                eventPayload[2].DataPointer = (IntPtr)receiveLinkStatePtr;
+
+                // bool maps to "win:Boolean", a 4-byte boolean
+                var isSessionReceiverInt = isSessionReceiver ? bool.True : bool.False;
```
```suggestion
                var isSessionReceiverInt = isSessionReceiver ? 1 : 0;
```

danielmarbach commented on sdk/servicebus/Azure.Messaging.ServiceBus/src/Diagnostics/ServiceBusEventSource.cs:
```diff
@@ -1509,7 +1592,34 @@ public void TransactionDischarged(string localTransactionId, ArraySegment<byte>
+                eventPayload[1].DataPointer = (IntPtr)amqpTransactionIdPtr;
+
+                // bool maps to "win:Boolean", a 4-byte boolean
+                var rollbackInt = rollback ? bool.True : bool.False;
```
```suggestion
                var rollbackInt = rollback ? 1 : 0;
```

danielmarbach committed:
Changing code online directly deserves to be punished

jsquire commented on sdk/servicebus/Azure.Messaging.ServiceBus/src/Diagnostics/ServiceBusEventSource.cs:
```diff
@@ -1509,7 +1592,34 @@ public void TransactionDischarged(string localTransactionId, ArraySegment<byte>
+                eventPayload[1].Size = (amqpTransactionId.Length + 1) * sizeof(char);
+                eventPayload[1].DataPointer = (IntPtr)amqpTransactionIdPtr;
+
+                // bool maps to "win:Boolean", a 4-byte boolean
```
Interesting; I'm assuming this has something to do with how we were marshalling the boolean and the expectations downstream.  Can you help me understand a bit more context?

danielmarbach replied:
My understanding is that due to alignment a bool needs up to four bytes. But since I used `Unsafe.SizeOf<bool>` as the size, which returns 1for the size, but the pointer points to a bool that gets aligned to four bytes this causes problem when reading it. This then crashes with an argument exception

jsquire commented:
/azp run net - servicebus - tests

jsquire commented:
.... and I just realized Josh already ran the tests, so now we get two clean runs.  😄

danielmarbach commented:
Ping ;)

JoshLove-msft commented:
> Ping ;)

Will take a look today - thanks again for your patience @danielmarbach!

JoshLove-msft reviewed (approved):
Thanks @danielmarbach!
